packages/eslint-plugin/src/rules/no-unsafe-enum-assignment.ts

```diff
@@ -0,0 +1,336 @@
+    ): boolean {
+      // First, recursively check for functions with type containers like:
+      //
+      // ```ts
+      // declare function useFruits(fruits: Fruit[]);
+      // useFruits([0, 1]);
+      // ```
+      if (util.isTypeReferenceType(argumentType)) {
+        const argumentTypeArguments =
+          typeChecker.getTypeArguments(argumentType);
+
+        const parameterSubTypes = tsutils.unionTypeParts(parameterType);
+        for (const parameterSubType of parameterSubTypes) {
+          if (!util.isTypeReferenceType(parameterSubType)) {
+            continue;
+          }
+          const parameterTypeArguments =
+            typeChecker.getTypeArguments(parameterSubType);
+
+          for (let i = 0; i < argumentTypeArguments.length; i++) {
+            if (
+              isMismatchedEnumFunctionArgument(
+                argumentTypeArguments[i],
+                parameterTypeArguments[i],
+              )
+            ) {
+              return true;
+            }
+          }
+        }
+
+        return false;
+      }
+
+      // Allow function calls that have nothing to do with enums, like:
+      //
+      // ```ts
+      // declare function useNumber(num: number);
+      // useNumber(0);
+      // ```
+      const parameterEnumTypes = getEnumTypes(typeChecker, parameterType);
+      if (parameterEnumTypes.length === 0) {
+        return false;
+      }
+
+      // Allow passing enum values into functions that take in the "any" type
+      // and similar types that should basically match any enum, like:
+      //
+      // ```ts
+      // declare function useNumber(num: number);
+      // useNumber(Fruit.Apple);
+      // ```
+      const parameterSubTypes = new Set(tsutils.unionTypeParts(parameterType));
+      for (const parameterSubType of parameterSubTypes) {
+        if (
+          util.isTypeFlagSet(
+            parameterSubType,
+            ALLOWED_TYPES_FOR_ANY_ENUM_ARGUMENT,
+          )
+        ) {
+          return false;
+        }
+      }
+
+      // Disallow passing number literals into enum parameters, like:
+      //
+      // ```ts
+      // declare function useFruit(fruit: Fruit);
+      // declare const fruit: Fruit.Apple | 1;
+      // useFruit(fruit)
+      // ```
+      return tsutils.unionTypeParts(argumentType).some(
+        argumentSubType =>
+          argumentSubType.isLiteral() &&
+          !util.isTypeFlagSet(argumentSubType, ts.TypeFlags.EnumLiteral) &&
+          // Permit the argument if it's a number the parameter allows, like:
+          //
+          // ```ts
+          // declare function useFruit(fruit: Fruit | -1);
+          // useFruit(-1)
+          // ```
+          // that's ok too
+          !parameterSubTypes.has(argumentSubType),
+      );
+    }
+
+    /**
+     * Checks whether a provided node mismatches
+     *
+     * @param provided
+     * @param recipient
+     */
+    function compareProvidedNode(
+      provided: TSESTree.Node,
```

Comment:
nit - in other parts of the codebase I have used "sender" and "receiver".
IDK if we want to standardise or not.
I was never happy about the names tho.

Reply:
Yeah I'm dissatisfied overall - and wonder if there's a language/industry-general term... will change here for standardization.

Reply:
Yeah I'm dissatisfied overall - and wonder if there's a language/industry-general term... TypeScript source often calls one _"target"_ but that feels more abstract to me. 🤷 

If you don't mind, I do like _"provided"_ and _"recipient"_ a bit more.